src/subject/messages.rs

```diff
@@ -1,54 +1,88 @@
+use crate::database::Executor;
 use actix_web::HttpResponse;
 use async_trait::async_trait;
 use serde::{Deserialize, Serialize};
 
-use super::model::{Subjects, SubjectsError};
 use crate::database::Connection;
 use crate::message::MessageResponder;
 
 #[derive(Deserialize, Serialize)]
 #[serde(tag = "type", content = "payload")]
 pub enum SubjectsMessage {
-    SubjectsQuery(SubjectsQuery),
+    SubjectsQuery(Option<serde_json::Value>),
 }
 
 #[async_trait]
 impl MessageResponder for SubjectsMessage {
     #[allow(clippy::async_yields_async)]
     async fn handle(&self, connection: Connection<'_, '_>) -> HttpResponse {
         match self {
-            SubjectsMessage::SubjectsQuery(message) => message.handle(connection).await,
+            SubjectsMessage::SubjectsQuery(_) => subjects_query(connection).await,
         }
     }
 }
 
-#[derive(Deserialize, Serialize)]
-#[serde(rename_all = "camelCase")]
-pub struct SubjectsQuery {}
-
-#[async_trait]
-impl MessageResponder for SubjectsQuery {
-    #[allow(clippy::async_yields_async)]
-    async fn handle(&self, connection: Connection<'_, '_>) -> HttpResponse {
-        let subjects = match connection {
-            Connection::Pool(pool) => Subjects::fetch(pool).await,
-            Connection::Transaction(transaction) => {
-                Subjects::fetch_via_transaction(transaction).await
-            }
-        };
-        match subjects {
-            Ok(data) => HttpResponse::Ok()
-                .content_type("application/json; charset=utf-8")
-                .json(&data),
-            Err(e) => {
-                println!("/subjects: {:?}", e);
-                match e {
-                    SubjectsError::DatabaseError { .. } => {
-                        HttpResponse::InternalServerError().finish()
-                    }
-                    SubjectsError::InvalidInstance => HttpResponse::InternalServerError().finish(),
-                }
-            }
+async fn subjects_query(connection: Connection<'_, '_>) -> HttpResponse {
+    let subjects = match connection {
+        Connection::Pool(pool) => fetch_subjects(pool).await,
+        Connection::Transaction(transaction) => fetch_subjects(transaction).await,
+    };
+    match subjects {
+        Ok(data) => HttpResponse::Ok()
+            .content_type("application/json; charset=utf-8")
+            .json(&data),
+        Err(e) => {
+            println!("/subjects: {:?}", e);
+            HttpResponse::InternalServerError().finish()
         }
     }
 }
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Subjects {
+    pub subjects: Vec<Subject>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Subject {
+    instance: String,
+    taxonomy_term_id: i32,
+}
+
+async fn fetch_subjects<'a, E>(executor: E) -> Result<Subjects, sqlx::Error>
```

Comment:
How did you get rid of `SubjectsError::InvalidInstance`?

Reply:
We fetch all subjects of all instances and make the filtering later in the API -> so we need no instance filter here any more